Tyin Sync

mausworks opened the pull request:
This PR adds the Tyin Sync plugin. 

It is used for syncing data with an external source; either by "pushing" or "pulling": A mechanism most developers are familiar with. Isn't "store" a synonym for "repository" anyways?

The plugin provides:
- A convenient way to get data in and out of your stores
- Request deduplication
- (Optional) request caching

Here's an example usage:

```tsx
import storeHook from "tyin/hook";
import sync from "tyin/plugin-sync";
import extend from "tyin/extend";
import { Note } from "@/types";

const useUserNotes = extend(storeHook<Note[]>([]))
  .with(
    sync({
      push: (notes, userId: string) =>
        fetch(`/api/notes/${userId}`, {
          method: "PUT",
          body: JSON.stringify(notes),
        }),
      pullOptions: { cacheDuration: 5000 },
      pull: (userId: string) =>
        fetch(`/api/notes/${userId}`).then((res) => res.json()),
    })
  )
  .seal();
```

There is no `onError`/`onSuccess`/`onResolved` handler, because that can just as easily be set up with `Promise.catch`, `then`, and `finally`.

mausworks committed:
fix: move react to devDependencies

If you're using Tying and not using React,
you're probably just not using React,
and thus won't use `storeHook`.
Otherwise, you're already using React,
in which case you can.

mausworks committed:
fix: move debounce to util-debounce

BREAKING CHANGE

This will be the pattern for all utils moving forward.

mausworks committed:
feat: add throttle utility

Similar to debounce, so it lives under util-

mausworks committed:
feat: add utility function for pruning objects

mausworks committed:
feat: add cache utility

mausworks committed:
fix: rename ObjectStoreOptions to ObjectAPIOptions

BREAKING CHANGE

technically

mausworks committed:
feat: support "deep plugins"

Plugins that return plugins

mausworks committed:
chore: move tests to __test__

mausworks committed:
feat: add Tyin Sync

Push and pull data in and out of your stores!

mausworks committed:
feat: experimental server selector

No need to try to memoize anything when there's only ever one render.

mausworks committed:
chore: lockfile

mausworks committed:
docs: update README

mausworks committed:
fix: ignore all .js and .d.ts files

mausworks committed:
fix: remove onError

Tip: Handle with `Promise.catch` instead.

mausworks committed:
docs: better Tyin Sync docs + divide quickstart into sections

mausworks committed:
fix: add missing license field

mausworks committed:
fix: add description & keywords

mausworks committed:
fix: add repository

mausworks committed:
fix: add author and homepage

mausworks committed:
docs: improve naming guidelines

mausworks committed:
docs: add badges

mausworks committed:
docs: clarify examples

mausworks committed:
docs: clearer guidelines

mausworks commented on src/plugin-sync/usePull.ts:
```diff
@@ -0,0 +1,38 @@
+
+export default function usePull<
+  T extends AnyState,
+  P extends PullFunction<any>
```
Can be `PullFunction<T>`

mausworks commented on src/plugin-sync/index.ts:
```diff
@@ -0,0 +1,225 @@
+  T extends AnyState,
+  S extends SyncSetup<T, PullFunction<T>>
+>(
+  setup: S extends SyncSetup<T, infer P> ? SyncSetup<T, P> : S
```
This weird thing helps with documentation, without it, you don't get any documentation at all for `setup` in VSCode, which is not great, because `SyncSetup` is quite complex.

I guess that's because, technically, `S` isn't documented, but `SyncSetup<T, P>` is…

I would prefer to make this `sync<T, P>` instead, but we need the exact type of `S` to infer the type of `SyncFunctions`.

mausworks committed:
fix: include sync plugin

mausworks committed:
feat: measure sync plugin individually

mausworks committed:
fix: use PullFunction<T>

mausworks committed:
WIP: extra args for push and delete

mausworks committed:
fix: add optional parameter to push example

mausworks committed:
test: add unit tests for the sync plugin

mausworks committed:
chore: target specific bun types

mausworks committed:
chore: lockfile

mausworks committed:
feat: support shallow plugins and improve types

mausworks committed:
docs: more realistic example for plugin-sync

- Shows how to use deep plugins
- Shows how to use `pull` for both PUT and POST

mausworks committed:
chore: remove comma

mausworks committed:
fix: move plugin-sync to top level

You can still put related functionality in a folder with the same name

mausworks committed:
docs: current sizes

mausworks commented on src/plugin-sync.ts:
```diff
@@ -0,0 +1,249 @@
+import prune, { Pruned } from "./util-prune";
+
+/**
+ * A function with that pulls the state from an external sourc using some arguments.
```
Fix typo

mausworks commented on src/plugin-sync.ts:
```diff
@@ -0,0 +1,249 @@
+
+/**
+ * A function that pushes the state to an external source,
+ * optionally using some extra arguments.
```
Move to above line

mausworks commented on src/plugin-sync.ts:
```diff
@@ -0,0 +1,249 @@
+
+/**
+ * A function that deletes the state from an external source,
+ * optionally using some extra arguments.
```
Move to above line

mausworks committed:
fix: bad constraint for the pull function + docs

mausworks committed:
fix: clean up ALL js, d.ts and .js.map files

tengl commented on src/hook.ts:
```diff
@@ -37,14 +37,16 @@ export type StateSelectorHook<T> = {
+): StateSelectorHook<T> => {
+  const useSelectValue = (selector: StateSelector<any> = (state) => state) =>
+    selector(store.get());
+  const useSelectMemoizedValue = (
```
See [useSyncExternalStoreWithSelector](https://github.com/facebook/react/blob/c5b9375767e2c4102d7e5559d383523736f1c902/packages/use-sync-external-store/src/useSyncExternalStoreWithSelector.js) for a more "robust" implementation of sync with selectors. I haven't really looked at all the detains in what it does, it was just something I found and thought might interest you.

mausworks replied:
> I haven't really looked at all the detains in what it does, it was just something I found and thought might interest you.

Cheers for having a look into it. ❤️ 

I looked into this before publishing 1.0.0 (zustand uses it), and a bit more now. It's quite neat, but it's [a lot more code](https://github.com/facebook/react/blob/main/packages/use-sync-external-store/src/useSyncExternalStoreWithSelector.js) to cover [concerns](https://github.com/facebook/react/blob/main/packages/use-sync-external-store/src/useSyncExternalStoreWithSelector.js#L50-L53) that Tyin doesn't have (our stores are already shared across all instances/concurrent copies—they are global), it's also (unfortunately) an external dependency.

As for why `useSelectMemoizedValue` is here: It's _only_ for clientside rendering. For SSR (when `typeof window === "undefined"`), there is no re-rendering (by definition). So instead of going through the trouble of setting up a `ref` and calling `useSyncExternalStore`, we can confidently just pull out whatever value is in the store. This is the "effective code path" that `useSyncExternalStore(WithSelector)` will take regardless, but I figure I might as well bypass it entirely.

By the way: The way you use Tyin (and zustand) with SSR is with a HOC or similar, so that you can "preload" values into the stores before you render the page. It's a bit scary since the stores are global, and the server will re-use whatever state is left in them for the next render. The way you solve that is simple though: never preload anything conditionally.

mausworks replied:
To recap what I said (since the comments are gone).

- I had a look at `useSyncExternalStoreWithSelector` before releasing 1.0.0, and again now. 👍 
- It mostly [covers concerns that we don't have (stores are global, so instance sharing is by design)](https://github.com/facebook/react/blob/c5b9375767e2c4102d7e5559d383523736f1c902/packages/use-sync-external-store/src/useSyncExternalStoreWithSelector.js#L50-L53)
- Server-side rendering can just pull the value directly from the store (no need to go through the useSyncExternalStore) even (server-side doesn't re-render, and the values never change after preloading\*)

\* Using global stores with SSR is a feels precarious since the stores are not (automatically) cleared out between requests (they are global and, thus, initialized on startup). The way you solve this is by never _conditionally_ preloading anything into the stores, that way, the data is never stale.

tengl replied:
Great! I haven't done much SSR, but it makes sense. 

I guess if `selector` **could be**  `undefined`, the client side code would essentially work as the SSR code? Because `equals` is only useful if the `selector` is defined? If I understand the code correctly.  Kind of like this: 

```ts
let s =  selector ? select : store.get;
return useSyncExternalStore(store.subscribe, s, s);
```

tengl commented on src/util-cache.ts:
```diff
@@ -0,0 +1,106 @@
+/** A value that will expire after a certain amount of time. */
+export type Transient<T> = readonly [T, number];
```
Wouldn't it be clearer if this was a `{value: T, lifetime: number}` instead?

Sure, it's only for internal use, but still.

mausworks replied:
I just added another argument, see: https://github.com/mausworks/tyin/pull/10/commits/c312364f637ed4c8a601cdae91c93ab38032415d

It had multiple benefits.

tengl commented on src/util-debounce.ts:
````diff
@@ -0,0 +1,30 @@
+ * ```
+ */
+export default function debounce<T extends (...args: any[]) => void>(
+  timeout: number,
+  callback: T
````
nit: Callback is usually before timeout in similar functions, in e.g. `setTimeout` .

mausworks replied:
Yeah, I know. It's just, a bit... yuk. I think it makes the most sense to reason "Delay for X, then do Y", so the args should come in that order. It also generally formats better. But I will reconsider before merging, thanks for the feedback. :)

tengl replied:
> Yeah, I know. It's just, a bit... yuk. I think it makes the most sense to reason "Delay for X, then do Y", so the args should come in that order. It also generally formats better. But I will reconsider before merging, thanks for the feedback. :)

Copied your deleted comment from the mail.

I agree that the order looks nicer with the number first. Callbacks tend also to span several lines which makes the timeout argument easy to miss. The only argument for having `timeout` last is to be inline with JavaScript APIs, but we can agree that that is not the best role model.  As long as this is consistent within the library, I think your style is better. 

I got a deja vu of this conversation, I think we had it before, and the same arguments came up. :smile:

mausworks replied:
I actually wrote it the wrong way around yesterday—the way it is in `setTimeout`. 🤦 

So _I am_ going to change this before release. Aesthetics be damned—it's hard to argue with tradition.

It's unlikely that the awkward ordering for `setTimeout` and `setInterval` is going to go away from the native APIs, so I might as well stick with them.

Thanks for pointing this out, it would've (secretly) driven me nuts a year from now probably. 👍

tengl commented on src/util-throttle.ts:
````diff
@@ -0,0 +1,53 @@
+ * ```
+ */
+export default function throttle<T extends (...args: any[]) => void>(
+  duration: number,
+  callback: T
````
nit: As above.

tengl reviewed (commented):
Just a quick review. Looks great overall.

mausworks committed:
fix: bad type in example

mausworks committed:
fix: simplify cache API

I don't like supporting three args for a function,
but in this case I'll make an exception,
since all alternatives are uglier and perform worse.

Plus, this shaves 130 bytes from `plugin-sync`.

mausworks committed:
docs: remove extra word

mausworks committed:
docs: fix typo and word line breaks

mausworks committed:
fix: lifetime issue

mausworks committed:
docs: updated size estimates

mausworks commented:
> Just a quick review. Looks great overall.

Thanks a whole lot @tengl! I have applied some changes based on your feedback.

tengl commented:
> Thanks a whole lot @tengl! I have applied some changes based on your feedback.

Awesome! 

BTW! Did you remove some comments? Because I got notified by mail about some comments but now I can't see them.

mausworks commented:
> BTW! Did you remove some comments? Because I got notified by mail about some comments but now I can't see them.

I did, they were duplicates (somehow got posted twice).
All content should still be there under "Resolved" conversations. :)

EDIT: They are... actually… gone now… cheers GitHub! 👍  

I'll quickly recap what I said in respective comments.

tengl commented:
Ahh, I guess it was a GUI bug that showed the same comment twice.

mausworks committed:
feat: support custom sync functions

Define a pull function by prefixing it with `pull`,
otherwise it's considered a general sync function.

Full type support. This one definitely didn't drive me nuts.

mausworks committed:
feat: `useHydrate` to handle hydration effects.

Has options for fetching at various times.

mausworks committed:
fix: bad export

mausworks committed:
docs: improve documentation and example

mausworks committed:
docs: better sync example

mausworks committed:
fix: broken badges

mausworks committed:
feat: return an augmented promise from useHydrate

This means you can use it with `use` and, thus, `Suspense`. 🎉

Still a bit WIP, maybe.

mausworks committed:
feat: move hooks to root

They can be used outside of the sync plugin

mausworks committed:
fix: rename duration to cacheTime

+ improve docs

mausworks committed:
docs: simplify docs

mausworks committed:
fix: minor export issues

mausworks committed:
fix: change order of throttle & debounce args

mausworks committed:
docs: fix bad cache docs

mausworks committed:
fix: tidy up useSync

mausworks committed:
fix: functions that start with pull are pull functions

mausworks committed:
fix: documentation and type fixes

mausworks committed:
feat: rehydrate on args change

mausworks committed:
chore: fix whitespace
